osu.Game/Online/SignalRDerivedTypeWorkaroundJsonConverter.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+using System;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A type of <see cref="JsonConverter"/> that serializes a subset of types used in multiplayer/spectator communication that
+    /// derive from a known base type. This is a safe alternative to using <see cref="TypeNameHandling.Auto"/> or <see cref="TypeNameHandling.All"/>,
+    /// which are known to have security issues.
+    /// </summary>
+    public class SignalRDerivedTypeWorkaroundJsonConverter : JsonConverter
+    {
+        public override bool CanConvert(Type objectType) =>
+            SignalRUnionWorkaroundResolver.BASE_TYPES.Contains(objectType) ||
+            SignalRUnionWorkaroundResolver.DERIVED_TYPES.Contains(objectType);
+
+        public override object? ReadJson(JsonReader reader, Type objectType, object? o, JsonSerializer jsonSerializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+                return null;
+
+            JObject obj = JObject.Load(reader);
+
+            string type = (string)obj[@"$dtype"]!;
+
+            var resolvedType = SignalRUnionWorkaroundResolver.DERIVED_TYPES.Single(t => t.Name == type);
```

Comment:
What's the point of having `CanConvert()` allow base types if the base types aren't allowed here? Also, what's the point of this at all? Why not just instantiate the type passed in? It's already guaranteed to be in the base/derived type lists.

Reply:
We are manually serialising the type in the line above. The incoming type will always be the base (`abstract`) type and is not usable here.

As for why `CanConvert` requires it: I believe that's for the deserialise case (where the incoming type is of the base type).

Reply:
Right. Kinda dodgy but okay. T-minus how many days until we don't use JSON/MessagePack and use something that's tried-and-tested for us...